### PR TITLE
feat: #1630 Pagination access control privilege endpoint

### PR DESCRIPTION
**Pull request by ianliuwk1019**
- Adjust endpoint "/access_control_privileges" to "/access-control-privileges"
- Add pagination schema objects for pagination related classes used by endpoint.
- Add "page_params" query parameter for "/access-control-privileges?application_id" for pagination capability.
- Use "create_date" as default sorting.
- Add simple paginate abstract repository to provide pagination functionality for server/admin_management.
- Implement paginate repository in access_control_privilege_repository for getting delegated admins.
- access_control_privilege_service uses and return paged delegated admins from repository paged method.
- Client-Code gen for frontend admin_management api.
- Adjust frontend due to endpoint change.
- Fix affected tests.

### CODE REVIEW

File: server/admin_management/api/app/constants.py [hunk 1]
```diff
@@ -101,6 +75,50 @@ class PrivilegeChangeTypeEnum(str, Enum):
     REVOKE = "REVOKE"
     UPDATE = "UPDATE"
 
+class SortOrderEnum(str, Enum):
+    ASC = "asc"
+    DESC = "desc"
+
+class DelegatedAdminSortByEnum(str, Enum):
+    # Note: this is not the exact model column name, requires table column mapping.
+    CREATE_DATE = "create_date"  # first one is the default sort field
```
Comment (craigyu) [inline server/admin_management/api/app/constants.py:84]: Is this how we determine the default sort field? Would it make sense to define a DEFAULT_SORT_COLUMN constant

File: server/admin_management/api/app/main.py [hunk 2]
```diff
@@ -82,7 +81,7 @@ def custom_generate_unique_id(route: APIRouter):
 )
 app.include_router(
     router_access_control_privilege.router,
-    prefix=apiPrefix + "/access_control_privileges",
+    prefix=apiPrefix + "/access-control-privileges",
```
Comment (craigyu) [inline server/admin_management/api/app/main.py:84]: 😍

**Review by craigyu — CHANGES_REQUESTED** [review]
I'm having trouble running this in docker since pytest is not installed for runtime Docker containers.

Looks like we have a `from pytest import Session` in `server/admin_management/api/app/repositories/simple_paginate_repository.py`

Perhaps it should be `from sqlalchemy.orm import Session`


Log
```
2024-11-13 18:30:03 INFO:     Will watch for changes in these directories: ['/usr/src']
2024-11-13 18:30:03 INFO:     Uvicorn running on http://0.0.0.0:8001 (Press CTRL+C to quit)
2024-11-13 18:30:03 INFO:     Started reloader process [7] using WatchFiles
2024-11-13 18:30:04 Process SpawnProcess-1:
2024-11-13 18:30:04 Traceback (most recent call last):
2024-11-13 18:30:04   File "/usr/local/lib/python3.12/multiprocessing/process.py", line 314, in _bootstrap
2024-11-13 18:30:04     self.run()
2024-11-13 18:30:04   File "/usr/local/lib/python3.12/multiprocessing/process.py", line 108, in run
2024-11-13 18:30:04     self._target(*self._args, **self._kwargs)
2024-11-13 18:30:04   File "/usr/local/lib/python3.12/site-packages/uvicorn/_subprocess.py", line 78, in subprocess_started
2024-11-13 18:30:04     target(sockets=sockets)
2024-11-13 18:30:04   File "/usr/local/lib/python3.12/site-packages/uvicorn/server.py", line 65, in run
2024-11-13 18:30:04     return asyncio.run(self.serve(sockets=sockets))
2024-11-13 18:30:04            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
2024-11-13 18:30:04   File "/usr/local/lib/python3.12/asyncio/runners.py", line 194, in run
2024-11-13 18:30:04     return runner.run(main)
2024-11-13 18:30:04            ^^^^^^^^^^^^^^^^
2024-11-13 18:30:04   File "/usr/local/lib/python3.12/asyncio/runners.py", line 118, in run
2024-11-13 18:30:04     return self._loop.run_until_complete(task)
2024-11-13 18:30:04            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
2024-11-13 18:30:04   File "uvloop/loop.pyx", line 1518, in uvloop.loop.Loop.run_until_complete
2024-11-13 18:30:04   File "/usr/local/lib/python3.12/site-packages/uvicorn/server.py", line 69, in serve
2024-11-13 18:30:04     await self._serve(sockets)
2024-11-13 18:30:04   File "/usr/local/lib/python3.12/site-packages/uvicorn/server.py", line 76, in _serve
2024-11-13 18:30:04     config.load()
2024-11-13 18:30:04   File "/usr/local/lib/python3.12/site-packages/uvicorn/config.py", line 433, in load
2024-11-13 18:30:04     self.loaded_app = import_from_string(self.app)
2024-11-13 18:30:04                       ^^^^^^^^^^^^^^^^^^^^^^^^^^^^
2024-11-13 18:30:04   File "/usr/local/lib/python3.12/site-packages/uvicorn/importer.py", line 22, in import_from_string
2024-11-13 18:30:04     raise exc from None
2024-11-13 18:30:04   File "/usr/local/lib/python3.12/site-packages/uvicorn/importer.py", line 19, in import_from_string
2024-11-13 18:30:04     module = importlib.import_module(module_str)
2024-11-13 18:30:04              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
2024-11-13 18:30:04   File "/usr/local/lib/python3.12/importlib/__init__.py", line 90, in import_module
2024-11-13 18:30:04     return _bootstrap._gcd_import(name[level:], package, level)
2024-11-13 18:30:04            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
2024-11-13 18:30:04   File "<frozen importlib._bootstrap>", line 1387, in _gcd_import
2024-11-13 18:30:04   File "<frozen importlib._bootstrap>", line 1360, in _find_and_load
2024-11-13 18:30:04   File "<frozen importlib._bootstrap>", line 1331, in _find_and_load_unlocked
2024-11-13 18:30:04   File "<frozen importlib._bootstrap>", line 935, in _load_unlocked
2024-11-13 18:30:04   File "<frozen importlib._bootstrap_external>", line 995, in exec_module
2024-11-13 18:30:04   File "<frozen importlib._bootstrap>", line 488, in _call_with_frames_removed
2024-11-13 18:30:04   File "/usr/src/api/app/main.py", line 7, in <module>
2024-11-13 18:30:04     from api.app.routers import (router_access_control_privilege,
2024-11-13 18:30:04   File "/usr/src/api/app/routers/router_access_control_privilege.py", line 6, in <module>
2024-11-13 18:30:04     from api.app.routers.router_guards import (
2024-11-13 18:30:04   File "/usr/src/api/app/routers/router_guards.py", line 12, in <module>
2024-11-13 18:30:04     from api.app.routers.router_utils import (
2024-11-13 18:30:04   File "/usr/src/api/app/routers/router_utils.py", line 5, in <module>
2024-11-13 18:30:04     from api.app.services.access_control_privilege_service import \
2024-11-13 18:30:04   File "/usr/src/api/app/services/access_control_privilege_service.py", line 9, in <module>
2024-11-13 18:30:04     from api.app.repositories.access_control_privilege_repository import \
2024-11-13 18:30:04   File "/usr/src/api/app/repositories/access_control_privilege_repository.py", line 9, in <module>
2024-11-13 18:30:04     from api.app.repositories.simple_paginate_repository import \
2024-11-13 18:30:04   File "/usr/src/api/app/repositories/simple_paginate_repository.py", line 10, in <module>
2024-11-13 18:30:04     from pytest import Session
2024-11-13 18:30:04 ModuleNotFoundError: No module named 'pytest'
```

**Comment by ianliuwk1019** [comment]
> I'm having trouble running this in docker since pytest is not installed for runtime Docker containers.
> 
> Looks like we have a `from pytest import Session` in `server/admin_management/api/app/repositories/simple_paginate_repository.py`
> 
> Perhaps it should be `from sqlalchemy.orm import Session`
> 
> Log
> 
> ```
> 2024-11-13 18:30:03 INFO:     Will watch for changes in these directories: ['/usr/src']
> 2024-11-13 18:30:03 INFO:     Uvicorn running on http://0.0.0.0:8001 (Press CTRL+C to quit)
> 2024-11-13 18:30:03 INFO:     Started reloader process [7] using WatchFiles
> 2024-11-13 18:30:04 Process SpawnProcess-1:
> 2024-11-13 18:30:04 Traceback (most recent call last):
> 2024-11-13 18:30:04   File "/usr/local/lib/python3.12/multiprocessing/process.py", line 314, in _bootstrap
> 2024-11-13 18:30:04     self.run()
> 2024-11-13 18:30:04   File "/usr/local/lib/python3.12/multiprocessing/process.py", line 108, in run
> 2024-11-13 18:30:04     self._target(*self._args, **self._kwargs)
> 2024-11-13 18:30:04   File "/usr/local/lib/python3.12/site-packages/uvicorn/_subprocess.py", line 78, in subprocess_started
> 2024-11-13 18:30:04     target(sockets=sockets)
> 2024-11-13 18:30:04   File "/usr/local/lib/python3.12/site-packages/uvicorn/server.py", line 65, in run
> 2024-11-13 18:30:04     return asyncio.run(self.serve(sockets=sockets))
> 2024-11-13 18:30:04            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
> 2024-11-13 18:30:04   File "/usr/local/lib/python3.12/asyncio/runners.py", line 194, in run
> 2024-11-13 18:30:04     return runner.run(main)
> 2024-11-13 18:30:04            ^^^^^^^^^^^^^^^^
> 2024-11-13 18:30:04   File "/usr/local/lib/python3.12/asyncio/runners.py", line 118, in run
> 2024-11-13 18:30:04     return self._loop.run_until_complete(task)
> 2024-11-13 18:30:04            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
> 2024-11-13 18:30:04   File "uvloop/loop.pyx", line 1518, in uvloop.loop.Loop.run_until_complete
> 2024-11-13 18:30:04   File "/usr/local/lib/python3.12/site-packages/uvicorn/server.py", line 69, in serve
> 2024-11-13 18:30:04     await self._serve(sockets)
> 2024-11-13 18:30:04   File "/usr/local/lib/python3.12/site-packages/uvicorn/server.py", line 76, in _serve
> 2024-11-13 18:30:04     config.load()
> 2024-11-13 18:30:04   File "/usr/local/lib/python3.12/site-packages/uvicorn/config.py", line 433, in load
> 2024-11-13 18:30:04     self.loaded_app = import_from_string(self.app)
> 2024-11-13 18:30:04                       ^^^^^^^^^^^^^^^^^^^^^^^^^^^^
> 2024-11-13 18:30:04   File "/usr/local/lib/python3.12/site-packages/uvicorn/importer.py", line 22, in import_from_string
> 2024-11-13 18:30:04     raise exc from None
> 2024-11-13 18:30:04   File "/usr/local/lib/python3.12/site-packages/uvicorn/importer.py", line 19, in import_from_string
> 2024-11-13 18:30:04     module = importlib.import_module(module_str)
> 2024-11-13 18:30:04              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
> 2024-11-13 18:30:04   File "/usr/local/lib/python3.12/importlib/__init__.py", line 90, in import_module
> 2024-11-13 18:30:04     return _bootstrap._gcd_import(name[level:], package, level)
> 2024-11-13 18:30:04            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
> 2024-11-13 18:30:04   File "<frozen importlib._bootstrap>", line 1387, in _gcd_import
> 2024-11-13 18:30:04   File "<frozen importlib._bootstrap>", line 1360, in _find_and_load
> 2024-11-13 18:30:04   File "<frozen importlib._bootstrap>", line 1331, in _find_and_load_unlocked
> 2024-11-13 18:30:04   File "<frozen importlib._bootstrap>", line 935, in _load_unlocked
> 2024-11-13 18:30:04   File "<frozen importlib._bootstrap_external>", line 995, in exec_module
> 2024-11-13 18:30:04   File "<frozen importlib._bootstrap>", line 488, in _call_with_frames_removed
> 2024-11-13 18:30:04   File "/usr/src/api/app/main.py", line 7, in <module>
> 2024-11-13 18:30:04     from api.app.routers import (router_access_control_privilege,
> 2024-11-13 18:30:04   File "/usr/src/api/app/routers/router_access_control_privilege.py", line 6, in <module>
> 2024-11-13 18:30:04     from api.app.routers.router_guards import (
> 2024-11-13 18:30:04   File "/usr/src/api/app/routers/router_guards.py", line 12, in <module>
> 2024-11-13 18:30:04     from api.app.routers.router_utils import (
> 2024-11-13 18:30:04   File "/usr/src/api/app/routers/router_utils.py", line 5, in <module>
> 2024-11-13 18:30:04     from api.app.services.access_control_privilege_service import \
> 2024-11-13 18:30:04   File "/usr/src/api/app/services/access_control_privilege_service.py", line 9, in <module>
> 2024-11-13 18:30:04     from api.app.repositories.access_control_privilege_repository import \
> 2024-11-13 18:30:04   File "/usr/src/api/app/repositories/access_control_privilege_repository.py", line 9, in <module>
> 2024-11-13 18:30:04     from api.app.repositories.simple_paginate_repository import \
> 2024-11-13 18:30:04   File "/usr/src/api/app/repositories/simple_paginate_repository.py", line 10, in <module>
> 2024-11-13 18:30:04     from pytest import Session
> 2024-11-13 18:30:04 ModuleNotFoundError: No module named 'pytest'
> ```

It certainly shouldn't be from `pytest` should be `sqlalchemy`. But this is interesting, I ran many times in my local with just now again and was fine, don't know why my local does not catch this. Sometimes I don't understand Python, quite flexible and seems it is smart enough to correct this things.

File: server/admin_management/api/app/constants.py (hunk 1 above)
Comment (ianliuwk1019) [reply server/admin_management/api/app/constants.py]: Ah, no, I was only intended to add a comment to say that is being used as default. I removed the confusing comment.
The real default chosen is set at the Schema for the router here:
image

**Comment by craigyu** [comment]
> It certainly shouldn't be from `pytest` should be `sqlalchemy`. But this is interesting, I ran many times in my local with just now again and was fine, don't know why my local does not catch this. Sometimes I don't understand Python, quite flexible and seems it is smart enough to correct this things.

@ianliuwk1019 i think it's because when you run locally in your terminal the pytest package is installed, but when you run it in docker or in a test or prod environment it doesn't get installed since it's only a dev dependency

**Review by craigyu — APPROVED** [review]
Do you think we should re-generate client-code-gen after the sortOrder change? I'm not sure if it's affected

**Comment by ianliuwk1019** [comment]
> Do you think we should re-generate client-code-gen after the sortOrder change? I'm not sure if it's affected

Yes, it does not re-generate/update new objects, but I update the json (has one single difference only).

**Pull request merged**